### PR TITLE
fix(core): wipe secret-bearing unknown fields and replaced map values

### LINKED ISSUE #902: core: give callers a way to wipe secret-bearing UnknownFields and Map values
## Context

Spotted by the `/crypto-privacy-review` gate on #901, which closed the encode-realloc hygiene gap in `encode_map_partial` (#893). That fix keeps secret bytes to one exactly-sized buffer inside the encoder. The gap below is adjacent, pre-existing, and deliberately out of scope there.

## Gap

The partial codec hands out secret-bearing values that an out-of-crate caller has **no way to wipe**.

1. **`UnknownFields` has no zeroize path at all.** `decode_map_partial` is `pub` and re-exported from `crates/core/src/codec/mod.rs`, so a caller outside `crates/core` receives an `UnknownFields` whose `entries` hold raw canonical value bytes. `Value::zeroize_bytes` and the corresponding `Map` path are `pub(crate)` (`crates/core/src/codec/value.rs`), so the caller cannot reach them. The `UnknownFields` vectors are simply dropped un-zeroized.

2. **`Map::insert` returns the replaced `Value`** (`crates/core/src/codec/value.rs`). The rewrite flow is `known.insert("contentKey", new_key);` — ignoring the return value is the ergonomic default, and it drops the previous secret un-zeroized. Same leak class as #893, on the caller's side of the seam.

## Exposure today

**Latent.** There are currently no callers of `decode_map_partial` / `encode_map_partial` outside `crates/core` — only `crates/core/tests/`, `crates/core/examples/kat_gen.rs`, and the `mod.rs` re-export. This bites the moment a slice wires the unknown-field-preserving re-encode path into the seal/grant surfaces, which is the same trigger #893 called out: that is the point these types start carrying inline content keys and grant/owner/history seed material across the crate boundary.

## Suggested direction

Consistent with AGENTS.md security rule 7 — zeroize at the terminal owner, never in a callee on a caller-owned buffer — the fix is to give the terminal owner the ability it currently lacks:

- `impl Drop for UnknownFields` that zeroizes each preserved raw vector, or a public explicit wipe.
- A public wipe on `Map`, so a caller holding decoded known fields can clear them.
- For `Map::insert`, decide deliberately: either `#[must_use]` on the returned old value so ignoring it is a lint, or an `insert` that zeroizes the value it replaces (and a separate `replace` for callers that want it back).

Land this **before** the first out-of-crate caller feeds secret-bearing fields through the partial codec.

## Dependencies

Depends on #893 (merged via #901) — that PR establishes the buffer-hygiene shape this extends to the seam.

**Blocks #635 (conditionally).** As of 2026-07-31 there is still no out-of-crate caller of `decode_map_partial` / `encode_map_partial` — the only references are `crates/core/src/codec/`, `crates/core/tests/`, and `crates/core/examples/kat_gen.rs`, and #870 (PR #903) did not add one. #635 is the most likely first consumer: it re-wraps grant blobs verbatim for the committed set and re-seals the owner blob, history link, and ascent link, which is exactly the surface that would carry inline content keys and grant/owner/history seed material across the crate boundary.

**This is a tripwire issue.** Whoever lands the first out-of-crate partial-codec caller must either close this first or add an explicit `Depends on #902` edge to their slice — otherwise the leak ships silently, because nothing in CI fails when an `UnknownFields` is dropped un-zeroized.

**Blocked out of scope, filed with dependency edges** (from the review gates on PR #935): #936 covers the decoder's in-flight tree and the seal layer's own preserved unknowns, both still dropped un-zeroized on shipping paths; #937 covers redacting `Value`/`Map`/`Version` renderings, of which this issue's `UnknownFields` redaction is one part.

Part of #655

### PR DESCRIPTION
**Pull request by FSM1**
## What

`decode_map_partial` is re-exported from `crates/core`, so a caller outside the crate receives preserved unknown-field bytes and decoded known fields with no reachable way to scrub either. Both are verbatim copies of whatever the field carried — inline content keys and grant/owner/history seed material on the seal surfaces. PR #901 closed the encoder-side half of this and filed the seam half as #902.

- **`UnknownFields` preserves into `Zeroizing<Vec<u8>>`.** The wipe is type-enforced at drop rather than a step an out-of-crate holder has to remember, and there is deliberately no public wipe: a wiped set would re-encode to bytes the decoder rejects, so the only wipe is the one after which nothing follows.
- **`entries()` borrows as plain bytes** rather than handing out `&Zeroizing<…>`. A borrow cannot carry the wipe, and `Zeroizing`'s derived `Debug` forwards to the inner `Vec`, so the old signature was a hole straight through the redaction below.
- **`Map::zeroize_bytes` is now `pub`.** The caller holding the decoded known fields is their terminal owner and can scrub them; `decode_map_partial`'s rustdoc states that obligation.
- **`Map::insert` wipes the value it replaces instead of returning it.** Discarding the return was the ergonomic form and the leak. `Map::remove` is the documented path for a caller that wants ownership.
- **The decode loop wipes the transient tree for an unknown entry** — an unknown field is decoded *and* copied verbatim, and the decoded `Value` is a second copy that dies in the loop.
- **A scrub guard covers every exit that does not return the known fields** — a reject or a panic out of the caller's `is_known` — matching `seal`'s `ScrubOnDrop`/`ScrubOwned` convention instead of a wipe on the `Err` arm only.
- **`UnknownFields` renders redacted**, with field names bounded through the same `DisplayKey` cap the error surface already uses on writer-controlled keys.

Terminal-owner discipline, AGENTS.md rule 7, holds throughout: every wipe runs on a buffer the wiping frame owns and is about to drop. The input `bytes: &[u8]` is caller-owned and left untouched.

No wire-format change and no KAT regeneration — the `crates/core/kat` tree is byte-identical to `main`.

## Fail-closed symmetry

`encode_map_partial` now rejects an empty preserved value. Verbatim-splicing one would emit a map head promising an entry with nothing after it — bytes this codec's own decoder rejects as `truncated`. The check is release-active and returns `Err`, per AGENTS.md rule 8 and the `assert_children_unique` convention, and the state is not reachable through the public API today, so the guard is what keeps that true.

## `Map::insert` signature — merge sequencing

`Map::insert` changes from `-> Option<Value>` to `-> ()`. **It is not `#[must_use]`,** so this is source-compatible with every existing call site: `crates/engine/src/settings.rs` calls it at `:363`, `:367`, `:374`, `:375`, `:379`, `:386` plus test sites, and all of them discard the result. `cargo check --workspace --all-targets` is green, so nothing in the workspace consumed the returned value. #928 and #932 touch `settings.rs` concurrently; neither is affected, and no merge order is forced.

## Tests

Three new, each verified red against a deliberate break:

- `a_caller_can_wipe_what_it_decoded` — the out-of-crate seam. Asserts the decoded known field and the preserved raw value both carry the secret, then that `Map::zeroize_bytes` scrubs the content key **in the buffer that held it** — capacity retained, so a wipe that swapped in a fresh `Vec` fails. Red with `Value::zeroize_bytes`'s `Bytes` arm removed: `left: Some(Bytes([171, 171, …])) right: Some(Bytes([]))`.
- `a_failed_decode_wipes_what_it_already_lifted` — a truncated second entry; the first entry's secret is scrubbed before the error propagates. Red with the scrub guard removed, same shape.
- `merged_encode_rejects_an_empty_preserved_value` — the rule-8 guard, **verified red in a release build**: gated behind `cfg!(debug_assertions)` the encoder happily produced `[162, 97, 118, 2, 102, 102, 117, 116, 117, 114, 101]`, a two-entry map head whose second key `"future"` has no value.
- `debug_redacts_preserved_values` — pins the rendering and that a 200-char crafted field name cannot flood a log line.

## Gates

`cargo fmt --all --check`, `cargo clippy --workspace --all-targets -- -D warnings`, `cargo check --workspace --all-targets`, `cargo check -p cipherbox-wasm --target wasm32-unknown-unknown --all-targets`, `cargo test -p cipherbox-core`, `cargo test --release -p cipherbox-core`, `cargo test -p cipherbox-engine` — all green.

`/simplify`, `/security-review` and `/crypto-privacy-review` all ran on `git diff main...HEAD`. Findings folded in: the `entries()` bypass of the redaction, the unwind hole in the `Err`-path wipe, the missing test for that wipe, the rule-8 empty-raw guard, unbounded key names in `Debug`, and weak assertions in two tests. Two findings are out of scope and filed with dependency edges: the seal layer's own preserved unknowns and the decoder's in-flight tree are still dropped un-zeroized, and `Value`/`Map` still render secrets in full.

Deliberately not taken: a validate-only `Decoder::skip_value` for unknown entries. It would remove the second copy and the wipe, but a second walker that must accept exactly what `read_value` accepts is a fail-closed divergence risk that outweighs the allocation saving on a latent path.

Closes #902
Part of #655

## Summary by CodeRabbit

* **Security Enhancements**
  * Sensitive decoded and replaced values are now securely wiped from memory.
  * Debug output redacts preserved unknown values to help prevent accidental disclosure.
  * Callers can explicitly clear decoded data while preserving unrelated unknown-field data.

* **Bug Fixes**
  * Improved cleanup when partial decoding fails.
  * Empty preserved values are now correctly rejected as truncated input.

* **API Updates**
  * Updated map insertion behavior and added an explicit method for retrieving removed values.

### CODE REVIEW

**Comment by FSM1** [comment]
@coderabbitai full review

**Comment by FSM1** [comment]
@coderabbitai full review

File: crates/core/src/codec/value.rs [hunk 1]
```diff
@@ -232,16 +232,20 @@ impl Map {
...
+    /// never secret). The wipe for a caller holding decoded known fields — see
+    /// [`Value::zeroize_bytes`].
+    ///
+    /// Terminal: a wiped map still encodes, to zero-length fields a schema
+    /// decoder rejects, so nothing may re-encode from it afterwards.
+    pub fn zeroize_bytes(&mut self) {
         for (_, v) in &mut self.entries {
             v.zeroize_bytes();
         }
     }
```
[the remark this answers is not in the document]
Comment (FSM1) [reply crates/core/src/codec/value.rs:245]: Confirmed factually, but deferring: this is misuse resistance, not the fail-closed asymmetry rule 8 covers, and it is not reachable today. Folded into #936 as gap 3.

**A wiped `Map` does re-encode — and its own decoder accepts the result.**

```text
{"contentKey": h'ab…ab' (32B), "name": "secret-file.txt", "v": 2}
  known.zeroize_bytes();
  encode(&Value::Map(known))  -> a3 61 76 02 64 6e616d65 60 6a 636f6e74656e744b6579 40
  decode(that)                -> Ok({"v": 2, "name": "", "contentKey": h''})
```

`encode_map_partial` behaves the same. Byte-identical under `cargo test --release`. No secret survives into the output.

**Why not rule 8.** Rule 8 asks for a release-active encode-side `Err` wherever the decoder *hard-rejects* what the encoder can emit — otherwise a release build signs bytes its own decoder always refuses. That is not what happens here: `h''` and `""` are canonical det-CBOR and `decode` returns `Ok`, as above. The one wipe-adjacent state that genuinely does produce decoder-rejected bytes — an empty preserved raw value, which would splice a key with nothing after it — is already guarded release-active at `crates/core/src/codec/fields.rs:204-206` and pinned by `merged_encode_rejects_an_empty_preserved_value`. So the rule-8 half of this surface is closed; what is left is a caller-misuse hazard, which is a lower bar and a design change to `Map` rather than a review-response commit.

**Not reachable.** `encode_map_partial` / `decode_map_partial` have no caller in `crates/core/src` or `crates/engine` — only tests, `kat_manifest.rs` and the `kat_gen` example; #902 is a latent seam. Every non-test `Map::zeroize_bytes` call site wipes and then drops (`crates/core/src/seal/body.rs:385,404`; `crates/core/src/codec/fields.rs:115,131`), and `encode_settings_body` encodes at `crates/engine/src/settings.rs:392` before it wipes at `:393`.

**Documented.** The obligation is stated at the wipe itself, `crates/core/src/codec/value.rs:238-239`:

> Terminal: a wiped map still encodes, to zero-length fields a schema decoder rejects, so nothing may re-encode from it afterwards.

One nuance carried into #936 rather than dropped: the "a schema decoder rejects" half only holds for fixed-length fields. `bytes_fixed::<N>` (`crates/core/src/seal/body.rs:455`) rejects `h''` with `InvalidFieldLength`, but `ChildRef::from_value` reads `name` via `as_text()` and `ipnsName` via `as_bytes()` with no length floor (`:90-91`), so a wiped `ChildRef` would re-encode and re-decode clean with both emptied. Silent data loss, not a loud reject — which is the part worth fixing, and why gap 3 keeps the terminal-flag sketch on file next to gap 1.
